File: apps/web/src/types/export-handoff.ts
import type { PublishTargetPlatform } from "./publish-draft";

export type ExportPackageStatus =
  | "DRAFT"
  | "READY_FOR_HANDOFF"
  | "HANDOFF_CREATED"
  | "FAILED_NEEDS_ATTENTION"
  | "CANCELLED";

export type PublishHandoffStatus =
  | "DRAFT"
  | "READY_FOR_OPERATOR"
  | "ACCEPTED"
  | "FAILED_NEEDS_ATTENTION"
  | "CANCELLED";

export type ReupQueueBatchAction =
  | "START_PROCESSING"
  | "START_AUTO_PIPELINE"
  | "SET_AUTOMATION"
  | "HOLD"
  | "RESUME"
  | "RETRY"
  | "CANCEL"
  | "MARK_MEDIA_READY"
  | "CREATE_EXPORT_PACKAGE"
  | "CREATE_PUBLISH_HANDOFF"
  | "DISMISS"
  | "PURGE";

export type BatchItemResult = {
  item_id: string;
  result: "succeeded" | "skipped" | "failed" | string;
  status: string | null;
  reason_code: string | null;
  message: string | null;
  export_package_id: string | null;
  publish_handoff_id: string | null;
};

export type BatchOperationResponse = {
  requested_count: number;
  succeeded_count: number;
  skipped_count: number;
  failed_count: number;
  export_package_id: string | null;
  publish_handoff_id: string | null;
  results: BatchItemResult[];
};

export type ReupQueueBatchActionRequest = {
  action: ReupQueueBatchAction;
  item_ids: string[];
  note?: string | null;
  target_platform?: PublishTargetPlatform | null;
  pipeline_mode?: string | null;
};

export type ExportPackageCreateRequest = {
  item_ids: string[];
  label?: string | null;
  operator_note?: string | null;
};

export type ExportPackageItem = {
  id: string;
  workspace_id: string;
  export_package_id: string;
  reup_queue_item_id: string;
  source_video_id: string;
  video_candidate_id: string;
  render_output_id: string | null;
  publish_draft_id: string | null;
  item_status: string;
  manifest_json: Record<string, unknown> | null;
  diagnostics_json: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export type ExportPackage = {
  id: string;
  workspace_id: string;
  status: ExportPackageStatus;
  label: string | null;
  operator_note: string | null;
  item_count: number;
  manifest_json: Record<string, unknown> | null;
  diagnostics_json: Record<string, unknown> | null;
  ready_at: string | null;
  failed_at: string | null;
  cancelled_at: string | null;
  items: ExportPackageItem[];
  publish_handoff_ids: string[];
  created_at: string;
  updated_at: string;
};

export type ExportPackageListResponse = {
  items: ExportPackage[];
  total_count: number;
  limit: number;
  offset: number;
};

export type PublishHandoffCreateRequest = {
  export_package_id: string;
  target_platform?: PublishTargetPlatform;
  operator_note?: string | null;
};

export type PublishHandoff = {
  id: string;
  workspace_id: string;
  export_package_id: string;
  target_platform: PublishTargetPlatform;
  status: PublishHandoffStatus;
  operator_note: string | null;
  payload_json: Record<string, unknown> | null;
  diagnostics_json: Record<string, unknown> | null;
  ready_at: string | null;
  accepted_at: string | null;
  failed_at: string | null;
  cancelled_at: string | null;
  created_at: string;
  updated_at: string;
};

export type PublishHandoffListResponse = {
  items: PublishHandoff[];
  total_count: number;
  limit: number;
  offset: number;
};

File: apps/api/src/schemas/export_handoff.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from src.enums import ExportPackageStatus, PublishHandoffStatus, PublishTargetPlatform, ReupQueueBatchAction, ReupQueueStatus


class BatchItemResultResponse(BaseModel):
    item_id: UUID
    result: str
    status: ReupQueueStatus | None = None
    reason_code: str | None = None
    message: str | None = None
    export_package_id: UUID | None = None
    publish_handoff_id: UUID | None = None


class BatchOperationResponse(BaseModel):
    requested_count: int
    succeeded_count: int
    skipped_count: int
    failed_count: int
    export_package_id: UUID | None = None
    publish_handoff_id: UUID | None = None
    results: list[BatchItemResultResponse]


class ExportPackageCreateRequest(BaseModel):
    item_ids: list[UUID] = Field(min_length=1, max_length=500)
    label: str | None = Field(default=None, max_length=180)
    operator_note: str | None = Field(default=None, max_length=1000)


class ExportPackageItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workspace_id: UUID
    export_package_id: UUID
    reup_queue_item_id: UUID
    source_video_id: UUID
    video_candidate_id: UUID
    render_output_id: UUID | None
    publish_draft_id: UUID | None
    item_status: str
    manifest_json: dict | None
    diagnostics_json: dict | None
    created_at: datetime
    updated_at: datetime


class ExportPackageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workspace_id: UUID
    status: ExportPackageStatus
    label: str | None
    operator_note: str | None
    item_count: int
    manifest_json: dict | None
    diagnostics_json: dict | None
    ready_at: datetime | None
    failed_at: datetime | None
    cancelled_at: datetime | None
    items: list[ExportPackageItemResponse] = Field(default_factory=list)
    publish_handoff_ids: list[UUID] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class ExportPackageListResponse(BaseModel):
    items: list[ExportPackageResponse]
    total_count: int
    limit: int
    offset: int


class PublishHandoffCreateRequest(BaseModel):
    export_package_id: UUID
    target_platform: PublishTargetPlatform = PublishTargetPlatform.FACEBOOK_REELS
    operator_note: str | None = Field(default=None, max_length=1000)


class PublishHandoffResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workspace_id: UUID
    export_package_id: UUID
    target_platform: PublishTargetPlatform
    status: PublishHandoffStatus
    operator_note: str | None
    payload_json: dict | None
    diagnostics_json: dict | None
    ready_at: datetime | None
    accepted_at: datetime | None
    failed_at: datetime | None
    cancelled_at: datetime | None
    created_at: datetime
    updated_at: datetime


class PublishHandoffListResponse(BaseModel):
    items: list[PublishHandoffResponse]
    total_count: int
    limit: int
    offset: int


class ReupQueueBatchActionRequest(BaseModel):
    action: ReupQueueBatchAction
    item_ids: list[UUID] = Field(min_length=1, max_length=500)
    note: str | None = Field(default=None, max_length=1000)
    target_platform: PublishTargetPlatform | None = None
    # START_AUTO_PIPELINE batch: auto_to_tts (default) | auto_to_render
    # SET_AUTOMATION batch: manual | auto_to_tts | auto_to_render (required)
    pipeline_mode: str | None = Field(default=None, max_length=40)
